LIPs/LIP-snapshot_earnings.md

```diff
@@ -0,0 +1,146 @@
+
+## Specification
+
+The snapshot will be referenced under the form of a Merkle tree where each bottom-level will be a leaf consisting of `{delegator, rewards, fees}`. The root hash of this tree will be stored on-chain and a delegator can submit its owed earnings along with a proof to the `BondingManager` contract to claim its past earnings up until and including the snapshot round. 
+
+### Merkle Tree Generation
+
+1. Establish the snapshot round
+
+2. Get all delegator addresses
+
+3. For each delegator call 
+    - `pendingStake(delegator, snapshotRound)`
+    - `pendingFees(delegator, snapShotRound)`
+
+4. Sort the tree by `delegator` address
+
+A library such as [merkletreejs](https://github.com/miguelmota/merkletreejs) can be used to generate the tree and validate proofs off-chain before submitting a transaction
+
+### Snapshot Round 
+
+#35 provides a clear transitional round after which claiming earnings are a constant operation (O(1)). 
+
+It would thus make sense to create the snapshot for the round prior to the upgrade round for #35. This will turn claiming earnings into a constant operation accross the board. 
+
+### `MerkleSnapshot` Contract
+
+The `MerkleSnapshot` contract is a generic contract that allows to verify merkle proofs for any state snapshots. 
+
+#### State
+
+##### `snapshots` mapping
+
+A map of `ID => merkleRoot` allowing the contract to support multiple snapshots in the future if need be. 
+
+`ID` in most cases is the hash of `LIP-X`. 
+
+e.g. for this LIP `keccak256("LIP-52")`
+
+#### API
+
+##### `snapshot(bytes32 _ID) returns (bytes32 merkleRoot)`
+
+Auto-generated `external` getter that gets the specific merkle tree root hash for an snapshot `_ID`.
+
+##### `setSnapshot(bytes32 _ID, bytes32 _merkleRoot)` 
+
+Sets the root hash of a merkle tree representing a snapshot and updates the `snapshots` map for `_ID` with `_merkleRoot`. 
+
+This function reverts if the map entry already exists. 
+
+
+##### `verify(bytes32 _ID, bytes32[] _proof, bytes32 _leaf) external returns (bool)`
+
+Calls `MerkleProof.verify()` with the `_proof` and `_leaf` to verify it against the merkle root for a specific snapshot `_ID`.
+
+```
+bytes32 rootHash = snapshots[_ID];
+MerkleProof.verify(_proof, rootHash, _leaf)
+```
+
+#### `MerkleProof` Library
+
+`MerkleEarnings` will use `MerkleProof` contract from OpezenZepplin  to verify proofs: https://docs.openzeppelin.com/contracts/3.x/api/cryptography#MerkleProof
+
+This function exposes a single internal API 
+
+```
+verify(bytes32[] memory proof, bytes32 root, bytes32 leaf) internal pure returns (bool)
+```
+
+### `BondingManager` API changes
+
+#### `claimSnapShotEarnings(uint256 _rewards, uint256 _fees, bytes32[] memory _earningsProof)`
+
+* `_rewards` is the amount of rewards owed to `msg.sender` up until and including the round the snapshot was taken. 
```

Comment:
If we're taking a snapshot of pendingStake then this should reflect rewards *plus* current stake. Maybe we could just say that the pendingStake value is used here (and call the param `_pendingStake`?).

Reply:
Fixed in b8872da